### PR TITLE
Cleanup: BrokerResponeJSON related cleanup.

### PR DESCRIPTION
**Pull request by mayankshriv**
Cleanup: Removed BrokerResponeJSON and replaced it with
BrokerResponse(Native).

1. BrokerResponseJSON is completely replaced by BrokerResponseNative.
2. All tests have been migrated to use BrokerResponseNative instead of
BrokerResponseJSON.
3. Removed DefaultReduceService class as it was used by
BrokerResponseJSON.

### CODE REVIEW

File: pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java [hunk 1]
```diff
@@ -149,7 +148,13 @@ private void runApproximationQueries(List<? extends AvroQueryGenerator.TestAggre
         if (query instanceof AvroQueryGenerator.TestSimpleAggreationQuery) {
           TestUtils.assertApproximation((Double) val, (Double) accurateValue, precision);
         } else {
-          TestUtils.assertJSONArrayApproximation((JSONArray) val, (JSONArray) accurateValue, precision);
+          List<GroupByResult> expectedValues = (List<GroupByResult>) accurateValue;
+          List<GroupByResult> groupByResults = (List<GroupByResult>) val;
+          for (int i = 0; i < groupByResults.size(); i++) {
+            double expected = Double.valueOf(groupByResults.get(i).getValue().toString());
+            Double actual = Double.valueOf(expectedValues.get(i).getValue().toString());
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java]: primitive double

File: pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java [hunk 2]
```diff
@@ -149,7 +148,13 @@ private void runApproximationQueries(List<? extends AvroQueryGenerator.TestAggre
...
+            TestUtils.assertApproximation(actual, expected, precision);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java]: If possible, use Assert.assertEquals(expected, actual, delta)

File: pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java [hunk 3]
```diff
@@ -230,16 +235,16 @@ public void testAggregation() throws Exception {
       final DataTable instanceResponse = QUERY_EXECUTOR.processQuery(instanceRequest);
       instanceResponseMap.clear();
       instanceResponseMap.put(new ServerInstance("localhost:0000"), instanceResponse);
-      final BrokerResponseJSON brokerResponse = REDUCE_SERVICE.reduceOnDataTable(brokerRequest, instanceResponseMap);
+      final BrokerResponseNative brokerResponse = REDUCE_SERVICE.reduceOnDataTable(brokerRequest, instanceResponseMap);
       LOGGER.info("BrokerResponse is " + brokerResponse.getAggregationResults().get(0));
       LOGGER.info("Result from avro is : " + aggCall.result);
       try {
-        Assert.assertEquals(Double.parseDouble(brokerResponse.getAggregationResults().get(0).getString("value")),
+        Assert.assertEquals(Double.parseDouble(brokerResponse.getAggregationResults().get(0).getValue().toString()),
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java]: Simply remove the try catch... (Maybe you want to add a delta?)

File: pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java [hunk 4]
```diff
@@ -260,33 +265,33 @@ public void testAggregationGroupBy() throws Exception {
...
       instanceResponseMap.clear();
       instanceResponseMap.put(new ServerInstance("localhost:0000"), instanceResponse);
-      final BrokerResponseJSON brokerResponse = REDUCE_SERVICE.reduceOnDataTable(brokerRequest, instanceResponseMap);
+      final BrokerResponseNative brokerResponse = REDUCE_SERVICE.reduceOnDataTable(brokerRequest, instanceResponseMap);
       LOGGER.info("BrokerResponse is " + brokerResponse.getAggregationResults().get(0));
       LOGGER.info("Result from avro is : " + groupBy.groupResults);
 
       try {
-        assertGroupByResults(brokerResponse.getAggregationResults().get(0).getJSONArray("groupByResult"),
+        assertGroupByResults(brokerResponse.getAggregationResults().get(0).getGroupByResult(),
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java]: Remove try catch

File: pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java [hunk 5]
```diff
@@ -389,13 +394,13 @@ public void testMatchAllQuery() throws RecognitionException, Exception {
...
-    final BrokerResponseJSON brokerResponse = REDUCE_SERVICE.reduceOnDataTable(brokerRequest, instanceResponseMap);
+    final BrokerResponseNative brokerResponse = REDUCE_SERVICE.reduceOnDataTable(brokerRequest, instanceResponseMap);
     LOGGER.info("BrokerResponse is " + brokerResponse);
     LOGGER.info("BrokerResponse is " + brokerResponse.getAggregationResults().get(0));
     LOGGER.info("BrokerResponse is " + brokerResponse.getAggregationResults().get(1));
 
-    Assert.assertEquals(brokerResponse.getAggregationResults().get(0).getInt("value"), 100000);
-    Assert.assertEquals(brokerResponse.getAggregationResults().get(1).getDouble("value"), 8.90662862E13);
+    Assert.assertEquals(Double.valueOf(brokerResponse.getAggregationResults().get(0).getValue().toString()), 100000.0);
+    Assert.assertEquals(Double.valueOf(brokerResponse.getAggregationResults().get(1).getValue().toString()), 8.90662862E13);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java]: Use parseDouble() for primitive double

File: pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java [hunk 6]
```diff
@@ -411,9 +416,9 @@ public void testRangeQuery() throws RecognitionException, Exception {
     final DataTable instanceResponse = QUERY_EXECUTOR.processQuery(instanceRequest);
     instanceResponseMap.clear();
     instanceResponseMap.put(new ServerInstance("localhost:0000"), instanceResponse);
-    final BrokerResponseJSON brokerResponse = REDUCE_SERVICE.reduceOnDataTable(brokerRequest, instanceResponseMap);
+    final BrokerResponseNative brokerResponse = REDUCE_SERVICE.reduceOnDataTable(brokerRequest, instanceResponseMap);
     LOGGER.info("BrokerResponse is " + brokerResponse.getAggregationResults().get(0));
-    Assert.assertEquals(brokerResponse.getAggregationResults().get(0).getInt("value"), 14);
+    Assert.assertEquals(Double.valueOf(brokerResponse.getAggregationResults().get(0).getValue().toString()), 14.0);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java]: parseDouble()

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByOperatorForMultiValueTest.java [hunk 7]
```diff
@@ -407,76 +409,73 @@ private BrokerResponseJSON getBrokerResponse(PlanMaker instancePlanMaker, Broker
...
-      Assert.assertEquals("[\"column7\",\"column6\"]", brokerResponse.getAggregationResults().get(i)
-          .getJSONArray("groupByColumns").toString());
-      Assert.assertEquals(15, brokerResponse.getAggregationResults().get(i).getJSONArray("groupByResult").length());
+      List<AggregationResult> aggregationResults = brokerResponse.getAggregationResults();
+      AggregationResult aggregationResult = aggregationResults.get(i);
+
+      List<String> groupByColumns = aggregationResult.getGroupByColumns();
+      Assert.assertTrue(groupByColumns.contains("column6"));
+      Assert.assertTrue(groupByColumns.contains("column7"));
+      Assert.assertTrue(groupByColumns.size() == 2);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByOperatorForMultiValueTest.java]: assertEquals()

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByOperatorForMultiValueTest.java [hunk 8]
```diff
@@ -407,76 +409,73 @@ private BrokerResponseJSON getBrokerResponse(PlanMaker instancePlanMaker, Broker
...
-        if ((i < 14 && aggResult[i] == aggResult[i + 1]) || (i > 0 && aggResult[i] == aggResult[i - 1])) {
-          //do nothing, as we have multiple groups within same value.
-        } else {
-          Assert.assertEquals(
-              groupResult[i],
-              brokerResponse.getAggregationResults().get(j).getJSONArray("groupByResult").getJSONObject(i)
-                  .getString("group"));
-        }
+        GroupByResult actualGroupByResult = actualAggregationResult.getGroupByResult().get(i);
+        double actual = Double.valueOf(actualGroupByResult.getValue().toString());
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByOperatorForMultiValueTest.java]: parseDouble()

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByOperatorTest.java [hunk 9]
```diff
@@ -412,68 +414,63 @@ public void testEmptyQueryResultsForInterSegmentAggregationGroupBy() throws Exce
...
-          //do nothing, as we have multiple groups within same value.
-        } else {
-          Assert.assertEquals(
-              groupResult[i],
-              brokerResponse.getAggregationResults().get(j).getJSONArray("groupByResult").getJSONObject(i)
-                  .getString("group"));
-        }
+        LOGGER.debug("Comparing group: {}", i);
+        GroupByResult actualGroupByResult = actualAggregationResult.getGroupByResult().get(i);
+        double actual = Double.valueOf(actualGroupByResult.getValue().toString());
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByOperatorTest.java]: parseDouble()

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByOperatorTest.java [hunk 10]
```diff
@@ -412,68 +414,63 @@ public void testEmptyQueryResultsForInterSegmentAggregationGroupBy() throws Exce
...
-  }
-
-  private void assertEmptyBrokerResponse(BrokerResponseJSON brokerResponse) throws JSONException {
+    Assert.assertEquals("count_star", brokerResponse.getAggregationResults().get(0).getFunction());
+    Assert.assertEquals("sum_met_impressionCount", brokerResponse.getAggregationResults().get(1).getFunction());
+    Assert.assertEquals("max_met_impressionCount", brokerResponse.getAggregationResults().get(2).getFunction());
+    Assert.assertEquals("min_met_impressionCount", brokerResponse.getAggregationResults().get(3).getFunction());
+    Assert.assertEquals("avg_met_impressionCount", brokerResponse.getAggregationResults().get(4).getFunction());
+    Assert.assertEquals("minMaxRange_met_impressionCount", brokerResponse.getAggregationResults().get(5).getFunction());
+    Assert.assertEquals("distinctCount_column12", brokerResponse.getAggregationResults().get(6).getFunction());
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByOperatorTest.java]: Change the order to actual, expected

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorMultiValueTest.java [hunk 11]
```diff
@@ -354,69 +353,62 @@ public void testEmptyQueryResultsForInterSegmentAggregationGroupBy() throws Exce
...
-      LOGGER.debug("For aggregation function: {}", _aggregationInfos.get(j));
-      final double[] aggResult = aggregationResult.get(j);
-      final String[] groupResult = groupByResult.get(j);
+      final double[] aggResult = aggregationResults.get(j);
+      final String[] expectedGroupByResult = groupByResult.get(j);
       for (int i = 0; i < 15; ++i) {
-        LOGGER.debug("Comparing group: {}", i);
+        AggregationResult aggregationResult = brokerResponse.getAggregationResults().get(j);
+        GroupByResult actualGroupByResult = aggregationResult.getGroupByResult().get(i);
         Assert.assertEquals(0, DoubleComparisonUtil.defaultDoubleCompare(aggResult[i],
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorMultiValueTest.java]: Is this equivalent to "Assert.assertEquals(Double.parseDouble(actualGroupByResult.getValue().toString()), aggResult[i])"?

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorMultiValueTest.java [hunk 12]
```diff
@@ -354,69 +353,62 @@ public void testEmptyQueryResultsForInterSegmentAggregationGroupBy() throws Exce
...
     }
-
   }
 
-  private void verifyResponse(BrokerResponseJSON brokerResponse, int groupLength)
-      throws JSONException {
-    Assert.assertEquals(_numAggregations, brokerResponse.getAggregationResults().size());
+  private void verifyResponse(BrokerResponseNative brokerResponse, int groupLength) {
+    List<AggregationResult> aggregationResults = brokerResponse.getAggregationResults();
+    Assert.assertEquals(_numAggregations, aggregationResults.size());
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorMultiValueTest.java]: actual, expected

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorTest.java [hunk 13]
```diff
@@ -361,61 +361,69 @@ public void testEmptyQueryResultsForInterSegmentAggregationGroupBy() throws Exce
...
     final List<String[]> groupByResult = getGroupResult();
     for (int j = 0; j < _numAggregations; ++j) {
-      final double[] aggResult = aggregationResult.get(j);
-      final String[] groupResult = groupByResult.get(j);
+      final double[] aggResult = aggregationResults.get(j);
+      final String[] expectedGroupByResult = groupByResult.get(j);
       for (int i = 0; i < 15; ++i) {
+        AggregationResult aggregationResult = brokerResponse.getAggregationResults().get(j);
+        GroupByResult actualGroupByResult = aggregationResult.getGroupByResult().get(i);
         Assert.assertEquals(0, DoubleComparisonUtil.defaultDoubleCompare(aggResult[i],
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorTest.java]: Assert.assertEquals(Double.parseDouble(actualGroupByResult.getValue().toString()), aggResult[i])?

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryOperatorForMultiValueTest.java [hunk 14]
```diff
@@ -403,67 +402,68 @@ public void testEmptyQueryResultsForInterSegmentAggregationGroupBy() throws Exce
...
+      final double[] expectedAggregationResult = expectedAggregationResults.get(j);
+      final String[] expectedGroupByResult = expectedGroupByResults.get(j);
       for (int i = 0; i < 15; ++i) {
         LOGGER.debug("Comparing group: {}", i);
-        Assert.assertEquals(0, DoubleComparisonUtil.defaultDoubleCompare(aggResult[i],
-            brokerResponse.getAggregationResults().get(j).getJSONArray("groupByResult").getJSONObject(i).getDouble("value")));
-        if (groupResult.length < 2) {
+        AggregationResult actualAggregationResult = brokerResponse.getAggregationResults().get(j);
+        GroupByResult actualGroupByResult = actualAggregationResult.getGroupByResult().get(i);
+        Assert.assertEquals(0, DoubleComparisonUtil.defaultDoubleCompare(expectedAggregationResult[i],
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryOperatorForMultiValueTest.java]: Assert.assertEquals(Double.parseDouble(actualGroupByResult.getValue().toString()), aggResult[i])?

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryOperatorTest.java [hunk 15]
```diff
@@ -433,64 +432,65 @@ public void testEmptyQueryResultsForInterSegmentAggregationGroupBy() throws Exce
...
+      final double[] expectedAggregationResult = expectedAggregationResults.get(j);
+      final String[] expectedGroupByResult = expectedGroupResults.get(j);
+
       for (int i = 0; i < 15; ++i) {
-        Assert.assertEquals(0, DoubleComparisonUtil.defaultDoubleCompare(aggResult[i],
-            brokerResponse.getAggregationResults().get(j).getJSONArray("groupByResult").getJSONObject(i).getDouble("value")));
-        if ((i < 14 && aggResult[i] == aggResult[i + 1]) || (i > 0 && aggResult[i] == aggResult[i - 1])) {
+        AggregationResult actualAggreagtionResult = brokerResponse.getAggregationResults().get(j);
+        GroupByResult actualGroupByResult = actualAggreagtionResult.getGroupByResult().get(i);
+        Assert.assertEquals(0, DoubleComparisonUtil.defaultDoubleCompare(expectedAggregationResult[i],
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryOperatorTest.java]: Assert.assertEquals(Double.parseDouble(actualGroupByResult.getValue().toString()), aggResult[i])

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationQueriesTest.java [hunk 16]
```diff
@@ -295,52 +295,50 @@ public void testInterSegmentAggregationFunctionPlanMakerAndRun() throws Exceptio
...
+    Assert.assertEquals(DoubleComparisonUtil.defaultDoubleCompare(6718981331945402400.0,
+        Double.parseDouble(aggregationResult.getValue().toString())), 0);
+
+    aggregationResult = brokerResponse.getAggregationResults().get(5);
+    Assert.assertEquals(aggregationResult.getFunction(), "distinctCount_column12");
+    Assert.assertEquals(Integer.parseInt(aggregationResult.getValue().toString()), 146);
+
+    aggregationResult = brokerResponse.getAggregationResults().get(6);
+    Assert.assertEquals(aggregationResult. getFunction(), "distinctCount_met_impressionCount");
+    Assert.assertEquals(Integer.parseInt(aggregationResult.getValue().toString()), 21);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationQueriesTest.java]: Change all defaultDoubleCompare to Assert.assertEquals(actual, expected, delta) for readability?

File: pinot-core/src/test/java/com/linkedin/pinot/query/plan/PlanMakerTest.java [hunk 17]
```diff
@@ -559,21 +559,21 @@ public void testInterSegmentSelectionPlanMaker()
...
-      LOGGER.debug(selectionResultsArray.getJSONArray(j).toString());
-      assertEquals(selectionResultsArray.getJSONArray(j).getInt(0), 1);
-      assertEquals(selectionResultsArray.getJSONArray(j).getInt(1), 91);
-      assertEquals(selectionResultsArray.getJSONArray(j).getInt(2) % 100, 91);
+    List<Serializable[]> rows = brokerResponse.getSelectionResults().getRows();
+    for (int i = 0; i < rows.size(); i++) {
+      Serializable[] actualValues = rows.get(i);
+      assertEquals(Double.valueOf(actualValues[0].toString()), 1.0);
+      assertEquals(Double.valueOf(actualValues[1].toString()), 91.0);
+      assertEquals(Double.valueOf(actualValues[2].toString()) % 100, 91.0);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/plan/PlanMakerTest.java]: parseValue

File: pinot-core/src/test/java/com/linkedin/pinot/query/plan/PlanMakerTest.java [hunk 18]
```diff
@@ -600,19 +600,19 @@ public void testInterSegmentSelectionNoOrderingPlanMaker() throws JSONException
...
     LOGGER.debug(brokerResponse.getSelectionResults().toString());
     LOGGER.debug("TimeUsedMs: {}", brokerResponse.getTimeUsedMs());
     LOGGER.debug(brokerResponse.toString());
 
-    JSONArray selectionResultsArray = brokerResponse.getSelectionResults().getJSONArray("results");
-    for (int j = 0; j < selectionResultsArray.length(); ++j) {
-      assertEquals(selectionResultsArray.getJSONArray(j).getInt(0), 1);
+    List<Serializable[]> rows = brokerResponse.getSelectionResults().getRows();
+    for (Serializable[] row : rows) {
+      Assert.assertEquals(Double.valueOf(row[0].toString()), 1.0);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/plan/PlanMakerTest.java]: parseDouble

**Comment by Jackie-Jiang** [comment]
LGTM otherwise. For double comparison, please try to use Assert.assertEquals(actual, expected, delta) for readability.

File: pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java (hunk 2 above)
Comment (mayankshriv) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/QueriesSentinelTest.java]: Seems like the too are not exactly same. Using TestUtils passes the test whereas using Assert does not. We can investigate the reason for that outside of this PR.

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorMultiValueTest.java (hunk 11 above)
Comment (mayankshriv) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorMultiValueTest.java]: Apparently not. Again, we can investigate why so, outside this PR.

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorTest.java (hunk 13 above)
Comment (mayankshriv) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryAndTrieTreeOperatorTest.java]: They are not the same.

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryOperatorForMultiValueTest.java (hunk 14 above)
Comment (mayankshriv) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationGroupByWithDictionaryOperatorForMultiValueTest.java]: They are not the same.

File: pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationQueriesTest.java (hunk 16 above)
Comment (mayankshriv) [inline pinot-core/src/test/java/com/linkedin/pinot/query/aggregation/AggregationQueriesTest.java]: The test fails on doing so.

**Pull request merged**